### ISSUE TITLE
Get selected file/folder in the explorer view

### ISSUE THREAD

**Issue by SamVerschueren** [Microsoft/vscode#3553]
It would be nice if it was possible to retrieve the selected file/folder in the explorer view. The use where I want to use it is for my `vscode-yo` extension to be able to run the generator in the selected directory. An API for this would be nice.

**Comment by bpasero** [comment]
@jrieken fyi

**Comment by stef-levesque** [comment]
This would be helpful for my [vscode-hexdump](https://marketplace.visualstudio.com/items?itemName=slevesque.vscode-hexdump) extension, to retrieve selected file not compatible with TextDocument (i.e. png, jpg, ...). I wouldn't need to show an inputbox to ask for the file path everything :)

**Comment by riteshatsencha** [comment]
Second that. It will help us for the extension we are developing @Sencha. [Here](http://stackoverflow.com/questions/38270751/vscode-api-for-file-explorer/38278890) is the link to Stack Overflow question on the same feature.

**Comment by fantacytyx** [comment]
Some extension need retrieve selected file/folder, such as [TortoiseSVN](https://marketplace.visualstudio.com/items?itemName=fantasytyx.tortoise-svn).
If can get  selected file/folder, even if  not integrate SVN, also can be very convenient use SVN by  contextmenu of explorer.

**Comment by jrieken** [comment]
May the [`menus`](https://code.visualstudio.com/docs/extensionAPI/extension-points#_contributesmenus) contribution we have added this summer helps anyone here. The contract is that when invoked from a resource (such as an explorer entry) its URI will be passed to the function

> When a command is invoked from a (context) menu, VS Code tries to infer the currently selected resource and passes that as a parameter when invoking the command. For instance, a menu item inside the Explorer is passed the URI of the selected resource and a menu item inside an editor is passed the URI of the document.

**Comment by fbricon** [comment]
@jrieken executing a command from the menu, I get the uri as expected, except it doesn't work when that command is invoked from a KB shortcut (I get an empty parameter instead).

**Comment by jrieken** [comment]
> except it doesn't work when that command is invoked from a KB shortcut

Yeah, thats the design of things as KB shortcuts are global (not contextual) by definition... It's like someone using you command-id with the `executeCommand`-api. The rule to always be prepared to be invoked without sufficient arguments. In those cases you need to figure them out by yourself. In this case by using an API that you don't have (yet)

**Comment by dalvikroid** [comment]
with some api  or not?

**Comment by alianZhen** [comment]
I find out a way to get the path of selected folder. You can contribute a menu of explorer,then run the vscode.commands.executeCommand('copyFilePath');. You can get the value into your clipboard

**Comment by akosyakov** [comment]
> In this case by using an API that you don't have (yet)

Do you have an idea for API? Something like `window.activeResource`, that returns a current selected resource of an active part?

**Comment by jrieken** [comment]
no

**Comment by matheus-sales** [comment]
I too need this function for a plugin. Has forecast for implementation?

**Comment by HenrikHolmIT** [comment]
I need this feature now. Has it ever been implemented or is there another way to get the selected items in the explorer view?

I need to search for specific files but only in the currently selected folder.

**Comment by smnbbrv** [comment]
I was surprised today this feature is not existing.

Would be very convenient to have a way to get a list of selected files / folders  as well.

**Comment by FuQiangZhang-Alex** [comment]
This feature is still not there.

**Comment by ctf0** [comment]
for the context menu u can do
- show on folder only
```jsonc
"menus": {
    "explorer/context": [
        {
            "when": "explorerResourceIsFolder",
            "command": "extension.hello_world",
            // ...
        }
    ]
}
```

- show on files only
```jsonc
"menus": {
    "explorer/context": [
        {
            "when": "!explorerResourceIsFolder",
            "command": "extension.hello_world",
            // ...
        }
    ]
}
```

and for the command
```js
vscode.commands.registerCommand('extension.hello_world', (e) => {
    let path = e.fsPath
    
    // do ur work
})
```

**Comment by HenrikHolmIT** [comment]
@ctf0 It does not. Extension code is activated by either from the command pallette or a shortcut. There is no interaction with the context menu in the explorer view.

Instead I try to find the currently active editor and trace back form that file.

**Comment by smurphy917** [comment]
Just to add my own use case here--I'm looking to enable users to select multiple files within the explorer, which my extension can then act on. As of right now, this doesn't appear to be possible without replicating the entire workspace folder structure in a custom TreeView (not even 100% sure it's possible then). I'd much prefer to just be able to retrieve the current list of selected files/folders on the OOTB explorer view.

**Comment by DragonFlux** [comment]
I would like to add my usecase as explained here: https://stackoverflow.com/questions/59909657/open-default-file-when-folder-is-selected-in-vscode

I believe we have a good variety of usecases now to back up this request!

**Comment by ngarg-panw** [comment]
I see that support for multi-file select is there in explorer view. In my extension i want to get the paths of all selected files. Right now, even if I select 5 files, i only get path for 1. I am not finding any relevant API. Do we have the API available for the same?

**Comment by commonsensesoftware** [comment]
To add to the party, I thought I'd add a related, but different use case. I am not and do not want to build an extension. I am trying to build a _Debug_ task when working with Rust. Rust will build a separate executable per integration test file. The debugging experience works just fine, but changing the **launch.json** file all the time to choose a different startup item doesn't make sense. The simplest bridge would seem to be to choose the currently selected file, but there is no such variable for this. I can make it work with the **${file}** variable, but only after opening the actual executable file and clicking through the _Do you really want to open this file_ prompt. That's workable, but very clunky.

Having an API for extension authors, including Rust, would be great, but it seems there is a use case for just having a variable that exposes that information is also useful. An API would certainly be more comprehensive for multi-select and so. Two new variables that simply indicate the currently selected folder and file would seem to be congruent with the existing **${selectedText}** variable.

**Comment by leojrobinovitch** [comment]
+1 for wanting this!

**Comment by rndware** [comment]
@ctf0 I managed to get your solution to work partially.

In my use case I don't want to have to right click a file or folder in the explorer view each time I want to run my extension. Instead  I want to use a simple keyboard shortcut.

I tried binding "extension.hello_world" to a keyboard shortcut in my keybindings.json:

```
    {
        "key": "ctrl+h",
        "command": "extension.hello_world",
        "when": "explorerViewletVisible && filesExplorerFocus"
    },
```

But if I'm focused on the explorer view and do <C+h> I find the following code:

```
vscode.commands.registerCommand('extension.hello_world', (e) => {
    let path = e.fsPath
    
    // do ur work
})

```
doesn't work as ```e``` comes back as 'undefined'.

**Comment by ctf0** [comment]
@rndware the api is for right click and chosing a cmnd from the context menu,

there is actually a way to use the same cmnd for both context menu and cmnd pallet while checking later where the value is comming from, thats what i use in https://github.com/ctf0/vscode-file-shortcut

**Comment by rndware** [comment]
@ctf0 I just debugged the vscode-file-shortcut extension.

The extension gets the path from the activeTextEdit.document when triggered from a keybinding because e is undefined:
```
            let doc = window.activeTextEditor.document
            let path = e ? e.fsPath : doc.uri.fsPath
```
Which won't be the correct behaviour if a user is highlighting a different item in the explorer view than the current active text document. e.g. a user presses up the up key and has not pressed enter yet.

**Comment by ArturoDent** [comment]
Using @alianZhen's idea from above https://github.com/microsoft/vscode/issues/3553#issuecomment-438969485 I implemented a *workaround* which seems to work well.  

See https://stackoverflow.com/a/65257167/836330 👍🏼 

```js
async function activate(context) {

   let createFile = vscode.commands.registerCommand('folder-operations.createFile', async (folder) => {

    // use this if triggered by a menu item,
    let newUri = folder;  // folder will be undefined when triggered by keybinding

     if (!folder) {                       // so triggered by a keybinding
       await vscode.commands.executeCommand('copyFilePath');
       folder = await vscode.env.clipboard.readText();  // returns a string

       // see note below for parsing multiple files/folders
       newUri = await vscode.Uri.file(folder);          // make it a Uri 
     }

     createFileOpen(newUri);                 // use in some function  
  });
  context.subscriptions.push(createFile);
}
```

see more notes on the SO answer.

**Comment by finnmerlett** [comment]
I would really like to see this implemented, would really help along the way to making it possible to get the full use out of VSCode with just the keyboard!

**Comment by bondiano** [comment]
Also will be cool to have path to selected file/folder in explorer as variable in Tasks like `${selectedFile}`/`selectedFolder`.

**Comment by moshfeu** [comment]
My extension needs that too. It allows a command only when `listDoubleSelection` but vscode still allow to set a keybinding.
When the user uses this shortcut, there is no option to get the selected resources.

https://github.com/moshfeu/vscode-compare-folders/issues/66

**Comment by QoobIY** [comment]
I think that if no one has done this in 5 years, we should do it yourself =)

**Comment by jdneo** [comment]
+1 for this.

When I try to adopt the new contribution point `file/newFile`, I wish I can create new files under the current selected folder in the file explorer.

**Comment by FlorianBuhl** [comment]
+1 for this.

**Comment by JeremyFunk** [comment]
A workaround that should work for most use cases (@ArturoDent version while preserving the clipboard state):

```typescript
async function activate(context) {

   let createFile = vscode.commands.registerCommand('folder-operations.createFile', async (folder) => {

    // use this if triggered by a menu item,
    let newUri = folder;  // folder will be undefined when triggered by keybinding

     if (!folder) {                       // so triggered by a keybinding
       const originalClipboard = await vscode.env.clipboard.readText();

       await vscode.commands.executeCommand('copyFilePath');
       folder = await vscode.env.clipboard.readText();  // returns a string

       await vscode.env.clipboard.writeText(originalClipboard);

       // see note below for parsing multiple files/folders
       newUri = await vscode.Uri.file(folder);          // make it a Uri 
     }

     createFileOpen(newUri);                 // use in some function  
  });
  context.subscriptions.push(createFile);
}
```
Using this for my hide files extension.

**Comment by tjx666** [comment]
Any progress on this api? I think this is a very basic api.

**Comment by R35007** [comment]
Yes I need that as well in part of my vscode. I need to get the list of files or folders selected by the user. Is there any update on this ticket ?

**Comment by eugeneniemand** [comment]
I found this solution. It wasn't very clear from docs that register command passes a URI which can then be used for getting the path of the selected file when activating your command from the context menu in the explorer
```typescript
export function activate(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "My-lite-switcher" is now active!');
	let disposable = vscode.commands.registerCommand('My-lite-switcher.activate', (uri:vscode.Uri) => {
		let f = uri.fsPath;
		vscode.window.showInformationMessage(`My Lite Switcher: ${f}`);
	});
	context.subscriptions.push(disposable);
}
```

**Comment by lramos15** [comment]
Does the new tabs API help anyone here? That's not limited by the text document and can get you the open file which is often the same as the selected file. I'm trying to understand the use cases in this thread and it seems like at least most would be solved with the tabs API.

**Comment by r3m0t** [comment]
I think they mean the file or folder which was clicked in the explorer view, or selected with the keyboard, not the open editor.

**Comment by dian-yu-luo** [comment]
> 

could I convert it to js

**Comment by Draugor** [comment]
> A workaround that should work for most use cases (@ArturoDent version while preserving the clipboard state):

this solution from @JeremyFunk is the closest i found to work, but there is still some situations where it doesn't work

- [x] right click, select menu item that runs command 
 (everything here after was not run via right-click menu)
- [x] select a file and open it, run command 
- [x] select multiple files, run command 
-> returns all
- [x] select folder, run command 
- [ ] have (one or) multiple files selected, but the currently opened file is not from this workspace (e.g. user/settings.json), run command 
-> returns the file path of the opened file (e.g. settings.json)
- [ ] select folder, click in editor window with a pre-opened file from the workspace (folder is still selected and not the file) 
-> returns the path of the file, not the selected folder
- [ ] select multiple files including the one currently opened, click in editor window with the opened file, run command 
-> returns only the path of the file in the editor window, not the other selected ones

**Comment by Simon-He95** [comment]
Indeed, this API is very basic and should be exposed. I hope vscode can support it.

**Comment by msweeney1999** [comment]
I need this too!  Any updates please?

**Comment by gravity-api** [comment]
This can be very easily achieved - please follow my post on StackOverflow
[https://stackoverflow.com/a/77659797/13547712](https://stackoverflow.com/a/77659797/13547712)

**Comment by msweeney1999** [comment]
I agree with [Simon-He95](https://github.com/Simon-He95).  Getting the URI for an Explorer Context menu **right** mouse click is already present in vscode.  What is needed is a URI notification for a **left** mouse click event or keyboard.

Something like this proposed API for a Treeview would be fantastic:

https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.treeViewActiveItem.d.ts#L24

The ```onDidChangeActiveItem``` event would specify the URI of the active item.

In fact, vscode's own git implementation could use this feature.  If I have a multi-rooted workspace, with subfolders containing repos, vscode git will always use the topmost folder as the git repo.   It should use the repo of the currently selected folder in Explorer.

**Comment by frudman** [comment]
adding my 2 cents: such an event emitter would be great to implement the most basic "Show README" for selected folder, a **very useful feature when browsing large repos**. This is exactly what GITHUB does whenever a folder is selected, the relevant README file is displayed. Would Love Love to get this same feature for vscode. I have the code ready for this (currently implemented as right-click context menu, but that's really a pita).

**Comment by ChiriVulpes** [comment]
I would love to be able to add keybindings for [these commands](https://marketplace.visualstudio.com/items?itemName=Aarilight.find-and-replace-in-selected-files) but can't because the extension can't currently support getting the current file list from the explorer. :<

**Comment by mirrortom** [comment]
Now, is there such an API? I need to use it too.
var selectedWorkspaceFolder=;
var fileOrDirs = selectedWorkspaceFolder.getSelectedItems();

**Comment by tgrue-openai** [comment]
Sadly the workaround where you copy clipboard, load path to clipboard, retrieve it, and restore state has two unacceptable issues for me.

1) Unsurprisingly, it's racey with the user. If they copy to the clipboard mid execution you can end up overwriting what they put in the clipboard. And, afaict, there is no solution to this.
2) The readText and writeText calls to the clipboard, at least on mac, lose rich state. This breaks some people's workflows and is particularly noticable.

Thus, add me to the bucket of people who are trying to show a README.md for a folder in my extension panel, with no way of being able to tell when a folder is selected by a left click.

**Comment by jenimanummic** [comment]
```ts
const command = commands.registerCommand("you.command", (file: Uri, selectedFiles: Uri[]) => {
	// your code here
});
context.subscriptions.push(command);
```

**Comment by tgrushka** [comment]
@jenimanummic does this actually work, or is it a suggestion to the VS Code authors? It would be nice if it worked as it could be suggested to the File Utils plugin author as a fix at https://github.com/sleistner/vscode-fileutils/issues/148.

> const command = commands.registerCommand("you.command", (file: Uri, selectedFiles: Uri[]) => {
> 	// your code here
> });
> context.subscriptions.push(command);

I, too, am trying to Super+D duplicate files in the file explorer on Linux. It seems to work with files (apparently, open files only), but does not work with folders ( 😦 ).

Thought I remember this working on macOS last time I used my Mac (maybe it's a quirk of macOS, no idea).

**Comment by Vasek-gh** [comment]
Maybe I'm missing something... but isn't this a basic thing that should work from the box? And it hasn't been implemented for at least 9 years!!! oh my god, 9 years!!! maybe even someone won't live to see the fix

**Comment by Vasek-gh** [comment]
@SamVerschueren can you please add workaround from @JeremyFunk to the issues description? I think this will save time for other people.
